Backport GMP changeset 4ac76064639e (2024-03-25).
printf: preserve embedded NUL characters by using vsprintf's return value.

Upstream: https://gmplib.org/repo/gmp/rev/4ac76064639e
Mailing-list patch: https://gmplib.org/list-archives/gmp-commit/2024-March.txt
Git mirror: https://github.com/gmp-mirror/gmp/commit/613987cf84c6acb740467f520ed9ec0770955e68

The native Windows runtime selects GMP's vsnprintf replacement. Without
this fix, MPFR 4.2.2 tfprintf/tsprintf check_null tests fail.

diff -r 7ecb3b2beea1 -r 1040c6303455 printf/doprntf.c
--- a/printf/doprntf.c	Sun Feb 18 20:20:57 2024 +0100
+++ b/printf/doprntf.c	Mon Mar 25 17:57:27 2024 +0100
@@ -267,8 +267,7 @@
 	 mean truncation */
       ASSERT (explen >= 0 && explen < sizeof(exponent)-1);
 #else
-      sprintf (exponent, p->expfmt, expsign, expval);
-      explen = strlen (exponent);
+      explen = sprintf (exponent, p->expfmt, expsign, expval);
       ASSERT (explen < sizeof(exponent));
 #endif
       TRACE (printf ("  expfmt %s gives %s\n", p->expfmt, exponent));
diff -r 7ecb3b2beea1 -r 1040c6303455 printf/repl-vsnprintf.c
--- a/printf/repl-vsnprintf.c	Sun Feb 18 20:20:57 2024 +0100
+++ b/printf/repl-vsnprintf.c	Mon Mar 25 17:57:27 2024 +0100
@@ -364,16 +364,14 @@
 
   if (total_width <= buf_size)
     {
-      vsprintf (buf, orig_fmt, orig_ap);
-      len = strlen (buf);
+      len = vsprintf (buf, orig_fmt, orig_ap);
     }
   else
     {
       char  *s;
 
       s = __GMP_ALLOCATE_FUNC_TYPE (total_width, char);
-      vsprintf (s, orig_fmt, orig_ap);
-      len = strlen (s);
+      len = vsprintf (s, orig_fmt, orig_ap);
       if (buf_size != 0)
 	{
 	  size_t  copylen = MIN (len, buf_size-1);
diff -r 7ecb3b2beea1 -r 1040c6303455 printf/sprintffuns.c
--- a/printf/sprintffuns.c	Sun Feb 18 20:20:57 2024 +0100
+++ b/printf/sprintffuns.c	Mon Mar 25 17:57:27 2024 +0100
@@ -53,9 +53,9 @@
 {
   char  *buf = *bufp;
   int   ret;
-  vsprintf (buf, fmt, ap);
-  ret = strlen (buf);
-  *bufp = buf + ret;
+  ret = vsprintf (buf, fmt, ap);
+  if (ret > 0)
+    *bufp = buf + ret;
   return ret;
 }
 
